cirq/google/known_devices.py

```diff
@@ -47,6 +52,50 @@ def _parse_device(s: str) -> Tuple[List[GridQubit], Dict[str, Set[GridQubit]]]:
     return qubits, measurement_lines
 
 
+def create_device_proto_from_diagram(
+        s: str,
+        gate_set: serializable_gate_set.SerializableGateSet = None,
+        durations: Dict[str, int] = None,
+) -> device_pb2.DeviceSpecification:
+    """
+    Parse ASCIIart device layout into DeviceSpecification proto containing
+    information about qubits and targets.  Note that this does not populate
+    the gate set information of the proto.
+
+    Also accepts a gate set object that will create the gate set proto.
+
+    This function also assumes that only adjacent qubits can be targets.
+    """
+    qubits, _ = _parse_device(s)
+    spec = device_pb2.DeviceSpecification()
+    grid_targets = spec.valid_targets.add()
+    grid_targets.name = TARGET_SET_NAME
+    grid_targets.target_ordering = device_pb2.TargetSet.SYMMETRIC
+    qubit_set = frozenset(qubits)
+    neighbor_set: Set[Tuple] = set()
+    spec.valid_qubits.extend([q.proto_id() for q in qubits])
+    for q in qubits:
+        for neighbor in [q + (0, 1), q + (1, 0), q + (-1, 0), q + (0, -1)]:
```

Comment:
Extract this into a method on `GridQubit`. Also on `LineQubit`.

Reply:
Done.